packages/presets/client/src/fragment-masking-plugin.ts

```diff
@@ -13,11 +24,44 @@ export type FragmentType<TDocumentType extends DocumentTypeDecoration<any, any>>
   : never;`;
 
 const makeFragmentDataHelper = `
+type UnionToIntersectGroupByTypeName<U, V = U> = [V] extends [
+  { __typename?: infer TypeName }
+]
+  ? TypeName extends any
+    ? UnionToIntersection<U extends { __typename?: TypeName } ? U : never>
+    : never
+  : never;
+
+type UnionFieldToIntersection<T> = [T] extends [never] ? never
+  : [T] extends [Array<unknown>]
+  ? Array<
+      | UnionFieldToIntersection<ExcludePrimitive<T[number]>>
+      | ExtractPrimitive<T[number]>
+    >
+  : UnionToIntersectGroupByTypeName<T> extends infer V
+  ? {
+      [Key in keyof V]:
+        | UnionFieldToIntersection<ExcludePrimitive<V[Key]>>
+        | ExtractPrimitive<V[Key]>;
+    }
+  : never;
+
+type Flatten<F> = [F] extends [never] ? never
+  : F extends Array<unknown>
+  ? Array<Flatten<ExcludePrimitive<F[number]>> | ExtractPrimitive<F[number]>>
+  : {
+      [Key in keyof Omit<F, " $fragmentRefs" | " $fragmentName">]:
+        | Flatten<ExcludePrimitive<F[Key]>>
+        | ExtractPrimitive<F[Key]>;
+    } & (F extends { " $fragmentRefs"?: { [K in string]: infer FRefs }; }
+      ? (FRefs extends any ? Flatten<FRefs> : never) : {}
+    );
+export type UnmaskFragment<F> = UnionFieldToIntersection<Flatten<F>>;
+
 export function makeFragmentData<
   F extends DocumentTypeDecoration<any, any>,
-  FT extends ResultOf<F>
->(data: FT, _fragment: F): FragmentType<F> {
-  return data as FragmentType<F>;
+>(data: UnmaskFragment<FragmentType<F>>, _fragment: F): FragmentType<F> {
+  return data as any;
```

Comment:
The main update of this PR is here.
`ResultOf<F>` does not accept any object if the fragment includes multiple fragments (I made an example #9702 ).
I made `UnmaskFragment` utility type to make it able to accept valid data for the fragment.